File: src/main/java/org/gtreimagined/gtlib/gui/SlotData.java
package org.gtreimagined.gtlib.gui;

import brachy.modularui.drawable.UITexture;
import brachy.modularui.widgets.slot.ModularSlot;
import lombok.Builder;
import lombok.Builder.Default;
import lombok.Getter;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.inventory.Slot;

@Builder
public class SlotData<T extends ModularSlot> {

    @Getter
    private SlotType<T> type;
    @Getter
    private int x;
    @Getter
    private int y;
    @Getter
    private int jeiX;
    @Getter
    private int jeiY;
    @Getter
    private UITexture baseTexture;
    @Getter
    private UITexture overlayTexture;
    @Getter
    @Default
    private int data = -1;

    public SlotData(SlotType<T> type, int x, int y, int jeiX, int jeiY, UITexture baseTexture, UITexture overlayTexture, int data){
        this.type = type;
        this.x = x;
        this.y = y;
        this.jeiX = jeiX;
        this.jeiY = jeiY;
        this.baseTexture = baseTexture;
        this.overlayTexture = overlayTexture;
        this.data = data;
    }

    public SlotData(SlotType<T> type, int x, int y, UITexture baseTexture, UITexture overlayTexture) {
        this.type = type;
        this.x = x;
        this.y = y;
        this.baseTexture = baseTexture;
        this.overlayTexture = overlayTexture;
    }

    public static <T extends ModularSlot> SlotDataBuilder<T> builder(){
        return new CustomSlotDataBuilder<>();
    }

    public static class CustomSlotDataBuilder<T extends ModularSlot> extends SlotDataBuilder<T> {
        boolean xSet = false, ySet = false, jeiXSet = false, jeiYSet = false, baseTextureSet = false, overlayTextureSet = false;

        @Override
        public SlotDataBuilder<T> x(int x) {
            xSet = true;
            return super.x(x);
        }
        @Override
        public SlotDataBuilder<T> y(int x) {
            ySet = true;
            return super.y(x);
        }

        @Override
        public SlotDataBuilder<T> jeiX(int jeiX) {
            jeiXSet = true;
            return super.jeiX(jeiX);
        }

        @Override
        public SlotDataBuilder<T> jeiY(int jeiY) {
            jeiYSet = true;
            return super.jeiY(jeiY);
        }

        @Override
        public SlotDataBuilder<T> baseTexture(UITexture baseTexture) {
            baseTextureSet = true;
            return super.baseTexture(baseTexture);
        }

        @Override
        public SlotDataBuilder<T> overlayTexture(UITexture overlayTexture) {
            overlayTextureSet = true;
            return super.overlayTexture(overlayTexture);
        }

        @Override
        public SlotData<T> build() {
            SlotData<T> slotData = super.build();
            if (slotData.getType() == null){
                throw new IllegalStateException("Slot Data must call type!");
            }
            if (!xSet){
                throw new IllegalStateException("X must be set");
            }
            if (!ySet){
                throw new IllegalStateException("Y must be set");
            }
            if (!jeiXSet) slotData.jeiX = slotData.x;
            if (!jeiYSet) slotData.jeiY = slotData.y;
            if (!baseTextureSet) slotData.baseTexture = slotData.type.getTexture();
            if (!overlayTextureSet) slotData.overlayTexture = slotData.type.getOverlay();
            return slotData;
        }
    }
}

File: src/main/java/org/gtreimagined/gtlib/gui/SlotData.kt
package org.gtreimagined.gtlib.gui

import brachy.modularui.api.drawable.IDrawable
import brachy.modularui.widgets.slot.ModularSlot
import java.util.function.Consumer

@JvmRecord
data class SlotData<T: ModularSlot>(
    val type: SlotType<T>,
    val x: Int,
    val y: Int,
    val jeiX: Int,
    val jeiY: Int,
    val background: IDrawable,
    val overlay: IDrawable,
    val slotGroup: Boolean,
    val data: Int
) {


    companion object {
        @JvmStatic
        fun <T: ModularSlot> create(consumer: Consumer<SlotDataBuilder<T>>): SlotData<T>{
            val builder = SlotDataBuilder<T>()
            consumer.accept(builder)
            val type = requireNotNull(builder.type) { "Must Specify a SlotType for SlotData"}
            val x = requireNotNull(builder.x) { "X must be set"}
            val y = requireNotNull(builder.y) { "Y must be set"}
            val jeiX = builder.jeiX ?: x
            val jeiY = builder.jeiY ?: y
            val baseTexture = builder.background ?: type.background
            val overlay = builder.overlay ?: type.overlay
            val slotGroup = builder.slotGroup ?: type.slotGroup
            return SlotData(type, x, y, jeiX, jeiY, baseTexture, overlay, slotGroup, builder.data)
        }
    }

    class SlotDataBuilder<T: ModularSlot> {
        var type: SlotType<T>? = null
        var x: Int? = null
        var y: Int? = null
        var jeiX: Int? = null
        var jeiY: Int? = null
        var background: IDrawable? = null
        var overlay: IDrawable? = null
        var slotGroup: Boolean? = null
        var data: Int = -1

        fun type(type: SlotType<T>): SlotDataBuilder<T>  = apply { this.type = type }
        fun x(x: Int): SlotDataBuilder<T> = apply { this.x = x }
        fun y(y: Int): SlotDataBuilder<T> = apply { this.y = y }
        fun jeiX(jeiX: Int): SlotDataBuilder<T> = apply { this.jeiX = jeiX }
        fun jeiY(jeiY: Int): SlotDataBuilder<T> = apply {this.jeiY = jeiY }
        fun background(baseTexture: IDrawable): SlotDataBuilder<T> = apply { this.background = baseTexture }
        fun overlay(overlayTexture: IDrawable): SlotDataBuilder<T> = apply { this.overlay = overlayTexture }
        fun slotGroup(slotGroup: Boolean): SlotDataBuilder<T> = apply { this.slotGroup = slotGroup }
        fun data(data: Int): SlotDataBuilder<T> = apply { this.data = data }
    }
}